src/secp256k1.c

```diff
@@ -316,6 +316,28 @@ int secp256k1_ec_pubkey_serialize(const secp256k1_context* ctx, unsigned char *o
     return ret;
 }
 
+int secp256k1_ec_pubkey_cmp(const secp256k1_context* ctx, const secp256k1_pubkey* pubkey0, const secp256k1_pubkey* pubkey1) {
+    unsigned char out[2][33];
+    size_t out_size = sizeof(out[0]);
+    const secp256k1_pubkey* pk[2];
+    int i;
+
+    VERIFY_CHECK(ctx != NULL);
+    pk[0] = pubkey0; pk[1] = pubkey1;
+    for (i = 0; i < 2; i++) {
+        /* A public key that is NULL or invalid is treated as being smaller than
+           any valid public key */
+        ARG_CHECK_NO_RETURN(pk[i] != NULL);
+        if (pk[i] == NULL
+                || !secp256k1_ec_pubkey_serialize(ctx, out[i], &out_size, pk[i], SECP256K1_EC_COMPRESSED)) {
+            out_size = sizeof(out[i]);
+            memset(out[i], 0, sizeof(out[i]));
+        }
+        VERIFY_CHECK(out_size == sizeof(out[i]));
+    }
+    return secp256k1_memcmp_var(out[0], out[1], sizeof(out[0]));
+}
```

Comment:
I think we should call the `illegal_callback` also if a pubkey is non-null but invalid. This will almost always crash the program, which is much cleaner than trying to include invalid keys in the order.

Independently of that question, I'm not sure if it's a good idea to continue reading user-provided values if we know that something is wrong in the code of the user, so we should return immediately when we see an invalid value (after the `illegal_callback`).

According to the description in the header, should the `illegal_callback` ever return (which is not recommended), then the return value of our function is undefined. So `ARG_CHECK` would be fine, even if it returns 0. If we still don't like the return value of 0 in that case, we could something like      

```suggestion
int secp256k1_ec_pubkey_cmp(const secp256k1_context* ctx, const secp256k1_pubkey* pubkey0, const secp256k1_pubkey* pubkey1) {
    unsigned char out[2][33];
    size_t out_size = sizeof(out[0]);
    const secp256k1_pubkey* pk[2];
    int i;

    VERIFY_CHECK(ctx != NULL);
    pk[0] = pubkey0; pk[1] = pubkey1;
    for (i = 0; i < 2; i++) {
        int valid_pk;
        /* A public key that is NULL or invalid is treated as being smaller than
           any valid or invalid public key */
        ARG_CHECK_NO_RETURN(pk[i] != NULL);
        if (pk == NULL) return (i == 1 ? -1 : 1);
         
        valid_pk = secp256k1_ec_pubkey_serialize(ctx, out[i], &out_size, pk[i], SECP256K1_EC_COMPRESSED);
        ARG_CHECK_NO_RETURN(valid_pubkey);
        if (!valid_pubkey) return (i == 1 ? -1 : 1);
        
        VERIFY_CHECK(out_size == sizeof(out[i]));
    }
    return secp256k1_memcmp_var(out[0], out[1], sizeof(out[0]));
}
```

This also has the advantage that we'll never fire the `illegal_callback` more than once. I don't think we promise that it will fire only once but so far we have adhered to that convention, which is helpful when writing tests.

Reply:
This is implicit as per the discussion in https://github.com/bitcoin-core/secp256k1/pull/783#issuecomment-695082220

> I think we should call the illegal_callback also if a pubkey is non-null but invalid.

`ec_pubkey_serialize` will do that through `pubkey_load`. Or are you suggesting to do that explicitly here?

> So ARG_CHECK would be fine, even if it returns 0

Yes, the lib doesn't promise a proper return value in that case but if you give an inconsistent comparator to a sorting algorithm, it may never terminate. The goal was therefore to give a consistent response even if the keys are invalid. As far as I can see, your suggested code change doesn't achieve that because for two invalid pubkeys A and B we have A < B and B < A.

Reply:
Resolved in https://github.com/jonasnick/secp256k1/tree/2020-11--pubkey-comparisons-jn